File: mcp/client/transports/websocket.py
"""WebSocket transport for MCP communication"""

import asyncio
import json
from typing import Dict, Optional

import aiohttp

from .base import BaseTransport
from ..types import JSONRPCRequest


class WebSocketTransport(BaseTransport):
    """
    WebSocket transport for MCP communication.
    Supports bidirectional streaming and notifications.
    """

    def __init__(
        self,
        endpoint: str,
        headers: Optional[Dict[str, str]] = None,
        timeout: int = 30000
    ):
        super().__init__(timeout)
        self.endpoint = endpoint
        self.headers = headers or {}
        self.session: Optional[aiohttp.ClientSession] = None
        self.ws: Optional[aiohttp.ClientWebSocketResponse] = None
        self._read_task: Optional[asyncio.Task] = None
        self.reconnect_attempts = 0
        self.max_reconnect_attempts = 5
        self.reconnect_delay = 1.0
        self.should_reconnect = True

    async def connect(self) -> None:
        """Connect WebSocket"""
        if self.connected:
            raise Exception("Already connected")

        await self._create_timeout_promise(
            self._connect_websocket(),
            "WebSocket connection"
        )

    async def _connect_websocket(self) -> None:
        """Internal WebSocket connection logic"""
        try:
            self.session = aiohttp.ClientSession()
            self.ws = await self.session.ws_connect(
                self.endpoint,
                headers=self.headers,
                timeout=aiohttp.ClientTimeout(total=self.timeout)
            )

            self.connected = True
            self.reconnect_attempts = 0

            # Start reading messages
            self._read_task = asyncio.create_task(self._read_messages())

        except Exception as e:
            if self.session:
                await self.session.close()
                self.session = None
            raise e

    async def disconnect(self) -> None:
        """Disconnect WebSocket"""
        if not self.connected:
            return

        self.should_reconnect = False

        try:
            # Cancel read task
            if self._read_task:
                self._read_task.cancel()
                try:
                    await self._read_task
                except asyncio.CancelledError:
                    pass

            # Close WebSocket
            if self.ws and not self.ws.closed:
                await self.ws.close()

            # Close session
            if self.session:
                await self.session.close()

            self.connected = False

        except Exception:
            pass

    async def send(self, message: JSONRPCRequest) -> None:
        """Send message via WebSocket"""
        if not self.connected or not self.ws:
            raise Exception("Not connected")

        try:
            await self.ws.send_json(message.__dict__)
        except Exception as e:
            raise Exception(f"Failed to send message: {e}")

    async def _read_messages(self) -> None:
        """Read messages from WebSocket"""
        try:
            if not self.ws:
                return

            async for msg in self.ws:
                if msg.type == aiohttp.WSMsgType.TEXT:
                    self._handle_message(msg.data)
                elif msg.type == aiohttp.WSMsgType.ERROR:
                    self._handle_error(Exception(f"WebSocket error: {self.ws.exception()}"))
                    break
                elif msg.type == aiohttp.WSMsgType.CLOSED:
                    break

        except asyncio.CancelledError:
            pass
        except Exception as e:
            self._handle_error(e)
        finally:
            self.connected = False
            self._handle_close()

            if self.should_reconnect and self.reconnect_attempts < self.max_reconnect_attempts:
                await self._reconnect()

    async def _reconnect(self) -> None:
        """Attempt to reconnect"""
        self.reconnect_attempts += 1
        delay = self.reconnect_delay * (2 ** (self.reconnect_attempts - 1))

        print(f"Reconnecting to WebSocket (attempt {self.reconnect_attempts}/{self.max_reconnect_attempts}) in {delay}s...")

        await asyncio.sleep(delay)

        try:
            await self._connect_websocket()
        except Exception as e:
            if self.reconnect_attempts >= self.max_reconnect_attempts:
                self._handle_error(Exception(f"Failed to reconnect after {self.max_reconnect_attempts} attempts"))

    def set_reconnect_options(self, max_attempts: int, delay: float) -> None:
        """Configure reconnection behavior"""
        self.max_reconnect_attempts = max_attempts
        self.reconnect_delay = delay

File: packages/mcp-client/src/transports/websocket.ts
import WebSocket from 'ws';

import { BaseTransport } from './base';
import { JSONRPCRequest } from '../types';

/**
 * WebSocket transport for MCP communication
 * Supports bidirectional streaming and notifications
 */
export class WebSocketTransport extends BaseTransport {
  private ws?: WebSocket;
  private endpoint: string;
  private headers: Record<string, string>;
  private reconnectAttempts: number = 0;
  private maxReconnectAttempts: number = 5;
  private reconnectDelay: number = 1000;
  private shouldReconnect: boolean = true;

  constructor(endpoint: string, headers: Record<string, string> = {}, timeout: number = 30000) {
    super(timeout);
    this.endpoint = endpoint;
    this.headers = headers;
  }

  async connect(): Promise<void> {
    if (this.connected) {
      throw new Error('Already connected');
    }

    return this.createTimeoutPromise(this.connectWebSocket(), 'WebSocket connection');
  }

  private connectWebSocket(): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      try {
        this.ws = new WebSocket(this.endpoint, {
          headers: this.headers,
        });

        this.ws.on('open', () => {
          this.connected = true;
          this.reconnectAttempts = 0;
          resolve();
        });

        this.ws.on('message', (data: WebSocket.Data) => {
          const message = data.toString();
          this.handleMessage(message);
        });

        this.ws.on('error', (error) => {
          this.handleError(error);
          if (!this.connected) {
            reject(error);
          }
        });

        this.ws.on('close', () => {
          this.connected = false;
          this.handleClose();

          if (this.shouldReconnect && this.reconnectAttempts < this.maxReconnectAttempts) {
            this.reconnect();
          }
        });
      } catch (error) {
        reject(error);
      }
    });
  }

  private async reconnect(): Promise<void> {
    this.reconnectAttempts++;
    const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);

    console.log(
      `Reconnecting to WebSocket (attempt ${this.reconnectAttempts}/${this.maxReconnectAttempts}) in ${delay}ms...`
    );

    await new Promise((resolve) => setTimeout(resolve, delay));

    try {
      await this.connectWebSocket();
    } catch (error) {
      if (this.reconnectAttempts >= this.maxReconnectAttempts) {
        this.handleError(
          new Error(`Failed to reconnect after ${this.maxReconnectAttempts} attempts`)
        );
      }
    }
  }

  async disconnect(): Promise<void> {
    if (!this.connected || !this.ws) {
      return;
    }

    this.shouldReconnect = false;

    return new Promise<void>((resolve) => {
      if (!this.ws) {
        resolve();
        return;
      }

      this.ws.once('close', () => {
        this.connected = false;
        resolve();
      });

      this.ws.close();

      // Force close after timeout
      setTimeout(() => {
        if (this.ws && this.ws.readyState !== WebSocket.CLOSED) {
          this.ws.terminate();
        }
        resolve();
      }, 5000);
    });
  }

  async send(message: JSONRPCRequest): Promise<void> {
    if (!this.connected || !this.ws) {
      throw new Error('Not connected');
    }

    return new Promise<void>((resolve, reject) => {
      if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
        reject(new Error('WebSocket not open'));
        return;
      }

      this.ws.send(JSON.stringify(message), (error) => {
        if (error) {
          reject(new Error(`Failed to send message: ${error.message}`));
        } else {
          resolve();
        }
      });
    });
  }

  setReconnectOptions(maxAttempts: number, delay: number): void {
    this.maxReconnectAttempts = maxAttempts;
    this.reconnectDelay = delay;
  }
}
